File: backend/src/domain/entities/StudySession.ts
export enum SessionStatus {
  ACTIVE = 'ACTIVE',
  PAUSED = 'PAUSED',
  COMPLETED = 'COMPLETED',
}

export class StudySession {
  constructor(
    public readonly id: string,
    public readonly userId: string,
    public subjectId: string | null,
    public semesterId: string | null,
    public activityName: string | null,
    public readonly startTime: Date,
    public endTime: Date | undefined,
    public pausedAt: Date | undefined,
    public status: SessionStatus,
    public totalDuration: number | undefined,
    public effectiveStudyTime: number | undefined,
    public breakCount: number,
    public accumulatedPauseTime: number,
    public readonly createdAt: Date,
    public updatedAt: Date
  ) {
    // Invariant: exactly one of (subjectId, activityName) is non-null.
    const hasSubject = subjectId !== null;
    const hasActivity = activityName !== null;
    if (hasSubject === hasActivity) {
      throw new Error(
        'StudySession must have exactly one of subjectId or activityName'
      );
    }
  }

  pause(): void {
    if (this.status !== SessionStatus.ACTIVE) {
      throw new Error('Can only pause an active session');
    }
    this.status = SessionStatus.PAUSED;
    this.pausedAt = new Date();
    this.updatedAt = new Date();
  }

  resume(): void {
    if (this.status !== SessionStatus.PAUSED) {
      throw new Error('Can only resume a paused session');
    }
    this.status = SessionStatus.ACTIVE;
    this.pausedAt = undefined;
    this.updatedAt = new Date();
  }

  stop(totalBreakTimeInSeconds: number = 0, totalPauseTimeInSeconds: number = 0): void {
    if (this.status === SessionStatus.COMPLETED) {
      throw new Error('Session already completed');
    }
    this.endTime = new Date();
    this.status = SessionStatus.COMPLETED;
    this.totalDuration = Math.floor((this.endTime.getTime() - this.startTime.getTime()) / 1000);
    this.effectiveStudyTime = this.totalDuration - totalBreakTimeInSeconds - totalPauseTimeInSeconds;
    this.updatedAt = new Date();
  }

  getCurrentDuration(): number {
    const endTime = this.endTime || new Date();
    return Math.floor((endTime.getTime() - this.startTime.getTime()) / 1000);
  }

  isActive(): boolean {
    return this.status === SessionStatus.ACTIVE;
  }

  belongsToUser(userId: string): boolean {
    return this.userId === userId;
  }

  isAdHoc(): boolean {
    return this.subjectId === null;
  }

  orphanToAdHoc(activityName: string): void {
    if (this.subjectId === null) {
      throw new Error('Session is already ad-hoc');
    }
    this.subjectId = null;
    this.semesterId = null;
    this.activityName = activityName;
    this.updatedAt = new Date();
  }

  static createForSubject(
    id: string,
    userId: string,
    subjectId: string,
    semesterId: string
  ): StudySession {
    return new StudySession(
      id,
      userId,
      subjectId,
      semesterId,
      null,
      new Date(),
      undefined,
      undefined,
      SessionStatus.ACTIVE,
      undefined,
      undefined,
      0,
      0,
      new Date(),
      new Date()
    );
  }

  static createAdHoc(
    id: string,
    userId: string,
    activityName: string
  ): StudySession {
    return new StudySession(
      id,
      userId,
      null,
      null,
      activityName,
      new Date(),
      undefined,
      undefined,
      SessionStatus.ACTIVE,
      undefined,
      undefined,
      0,
      0,
      new Date(),
      new Date()
    );
  }
}

File: mobile/lib/src/domain/models/session/study_session.dart
enum SessionStatus { active, paused, completed }

class StudySession {
  StudySession({
    required this.id,
    required this.subjectId,
    required this.activityName,
    required this.startTime,
    required this.status,
    required this.accumulatedPauseTime,
    required this.breakCount,
    this.endTime,
    this.pausedAt,
    this.totalDuration,
    this.effectiveStudyTime,
  }) : assert(
          (subjectId == null) != (activityName == null),
          'StudySession must have exactly one of subjectId or activityName',
        );

  factory StudySession.fromJson(Map<String, dynamic> json) => StudySession(
        id: json['id'] as String,
        subjectId: json['subjectId'] as String?,
        activityName: json['activityName'] as String?,
        startTime: DateTime.parse(json['startTime'] as String).toLocal(),
        endTime: json['endTime'] == null
            ? null
            : DateTime.parse(json['endTime'] as String).toLocal(),
        pausedAt: json['pausedAt'] == null
            ? null
            : DateTime.parse(json['pausedAt'] as String).toLocal(),
        status: _parseStatus(json['status']),
        accumulatedPauseTime: (json['accumulatedPauseTime'] as num?)?.toInt() ?? 0,
        breakCount: (json['breakCount'] as num?)?.toInt() ?? 0,
        totalDuration: (json['totalDuration'] as num?)?.toInt(),
        effectiveStudyTime: (json['effectiveStudyTime'] as num?)?.toInt(),
      );

  final String id;
  final String? subjectId;
  final String? activityName;
  final DateTime startTime;
  final DateTime? endTime;
  final DateTime? pausedAt;
  final SessionStatus status;
  final int accumulatedPauseTime; // seconds
  final int breakCount;
  final int? totalDuration; // seconds
  final int? effectiveStudyTime; // seconds

  /// True when this session has no subject — its label is `activityName` and
  /// it aggregates into the dashboard's "other" totals row.
  bool get isAdHoc => subjectId == null;

  /// What the UI shows in the chip slot / history row in place of a subject
  /// name. For subject sessions, the caller resolves the subject and renders
  /// its name; for ad-hoc, this is the typed activity name.
  String get adHocLabel => activityName ?? '';

  /// Wall-clock derived effective elapsed (seconds) — never persisted as a
  /// counter (ADR-0003). When active, ticks against `now`; when paused, frozen
  /// at the moment of pause; when completed, equal to `effectiveStudyTime`.
  int effectiveElapsedAt(DateTime now) {
    final reference = switch (status) {
      SessionStatus.active => now,
      SessionStatus.paused => pausedAt ?? now,
      SessionStatus.completed => endTime ?? now,
    };
    final wallClock =
        reference.difference(startTime).inSeconds - accumulatedPauseTime;
    return wallClock < 0 ? 0 : wallClock;
  }

  static SessionStatus _parseStatus(dynamic raw) {
    final value = (raw as String?)?.toUpperCase();
    return switch (value) {
      'ACTIVE' => SessionStatus.active,
      'PAUSED' => SessionStatus.paused,
      'COMPLETED' => SessionStatus.completed,
      _ => SessionStatus.completed,
    };
  }
}
